types/opt/value.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package opt
+
+import (
+	"fmt"
+	"reflect"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+)
+
+// Value is an optional value to be JSON-encoded.
+// Unlike the specialized Bool type, an empty (unset) Value[T] is marshaled as
+// a JSON null regardless of the "omitempty" JSON tag option when using [encoding/json].
+// With [github.com/go-json-experiment/json], however, empty optional Values are
+// omitted from the output.
+type Value[T any] struct {
+	value T
+	set   bool
+}
+
+// Equal reports whether the receiver and the other value are equal.
+// If the template type T in Value[T] implements an Equal method, it will be used
+// instead of the == operator for comparing values.
+type equatable[T any] interface {
+	// Equal reports whether the receiver and the other values are equal.
+	Equal(other T) bool
+}
+
+// Empty returns an empty optional Value.
+func Empty[T any]() Value[T] {
+	return Value[T]{set: false}
+}
+
+// Of returns an optional Value containing the specified value.
+// It treats nil pointers, unsafe pointers, and interfaces as unset values,
+// but nil slices and maps as empty slices and maps.
+func Of[T any](v T) Value[T] {
+	switch rv := reflect.ValueOf(v); rv.Kind() {
+	case reflect.Invalid:
+		// A nil interface (e.g. `any(nil)`) is equivalent to an unset value.
+		// It should be marshalled as null, or be omitted from the output.
+		return Empty[T]()
+	case reflect.Pointer, reflect.UnsafePointer, reflect.Interface:
+		if rv.IsNil() {
+			// A nil pointer, unsafe pointer, or interface is equivalent to an unset value.
+			// It should be marshalled as null, or be omitted from the output.
+			return Empty[T]()
+		}
+	case reflect.Map, reflect.Slice:
+		if rv.IsNil() {
+			// A nil map or slice is equivalent to an empty map or slice,
+			// and should be marshalled as `{}` or `[]`, accordingly.
+			return Value[T]{value: v, set: true}
+		}
+	}
+	return Value[T]{value: v, set: true}
+}
+
+// String implements Stringer.
+func (o *Value[T]) String() string {
+	if !o.set {
+		return fmt.Sprintf("(empty[%T])", o.value)
+	}
+	return fmt.Sprint(o.value)
+}
+
+// Set assigns the specified value to the optional value o.
+func (o *Value[T]) Set(v T) {
+	*o = Of(v)
+}
+
+// Clear resets o to an empty state.
+func (o *Value[T]) Clear() {
+	*o = Empty[T]()
+}
+
+// Present reports whether o has a value set.
+func (o *Value[T]) Present() bool {
```

Comment:
Optional: Could also call this `Has`.

Reply:
`Has` is shorter, but `Present` appears to be used more widely in Go. So I went with the current name.

Reply:
At least within the stdlib, `Has` is much more common:
```
$ rgrep "func.* Has[(A-Z]"
strings/strings.go:463:func HasPrefix(s, prefix string) bool {
strings/strings.go:468:func HasSuffix(s, suffix string) bool {
bytes/bytes.go:539:func HasPrefix(s, prefix []byte) bool {
bytes/bytes.go:544:func HasSuffix(s, suffix []byte) bool {
cmd/vendor/github.com/google/pprof/profile/profile.go:731:func (s *Sample) HasLabel(key, value string) bool {
cmd/vendor/github.com/google/pprof/profile/profile.go:825:func (p *Profile) HasFunctions() bool {
cmd/vendor/github.com/google/pprof/profile/profile.go:836:func (p *Profile) HasFileLines() bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:82:func (r *Config) HasProgram(s string) bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:86:func (r *Config) HasGOOS(s string) bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:90:func (r *Config) HasGOARCH(s string) bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:94:func (r *Config) HasGoVersion(s string) bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:98:func (r *Config) HasVersion(program, version string) bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:102:func (r *Config) HasCounter(program, counter string) bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:106:func (r *Config) HasCounterPrefix(program, prefix string) bool {
cmd/vendor/golang.org/x/telemetry/internal/config/config.go:110:func (r *Config) HasStack(program, stack string) bool {
cmd/vendor/golang.org/x/tools/go/analysis/passes/internal/analysisutil/util.go:30:func HasSideEffects(info *types.Info, e ast.Expr) bool {
cmd/vendor/golang.org/x/tools/internal/stdlib/stdlib.go:65:func HasPackage(path string) bool {
cmd/vendor/golang.org/x/tools/internal/typeparams/free.go:23:func (w *Free) Has(typ types.Type) (res bool) {
cmd/vendor/golang.org/x/text/internal/language/language.go:191:func (t Tag) HasString() bool {
cmd/vendor/golang.org/x/text/internal/language/language.go:275:func (t Tag) HasVariants() bool {
cmd/vendor/golang.org/x/text/internal/language/language.go:280:func (t Tag) HasExtensions() bool {
cmd/go/internal/modload/init.go:618:func HasModRoot() bool {
cmd/go/internal/str/path.go:16:func HasPathPrefix(s, prefix string) bool {
cmd/go/internal/str/path.go:38:func HasFilePathPrefix(s, prefix string) bool {
cmd/go/internal/toolchain/switch.go:226:func HasAuto() bool {
cmd/go/internal/toolchain/switch.go:232:func HasPath() bool {
cmd/compile/internal/inline/inl.go:95:func HasPgoHotInline(fn *ir.Func) bool {
cmd/compile/internal/base/hashdebug.go:137:func HasDebugHash() bool {
cmd/compile/internal/ir/node.go:418:func (s NameSet) Has(n *Name) bool {
cmd/compile/internal/ir/node.go:499:func HasUniquePos(n Node) bool {
cmd/compile/internal/ir/func.go:253:func (f *Func) HasDefer() bool                 { return f.flags&funcHasDefer != 0 }
cmd/compile/internal/ssa/_gen/main.go:416:	fmt.Fprintln(w, "func (o Op) HasSideEffects() bool { return opcodeTable[o].hasSideEffects }")
cmd/compile/internal/ssa/opGen.go:40875:func (o Op) HasSideEffects() bool { return opcodeTable[o].hasSideEffects }
cmd/compile/internal/syntax/type.go:54:func (f exprFlags) HasOk() bool           { return f&128 != 0 }
cmd/compile/internal/types2/api.go:433:func (tv TypeAndValue) HasOk() bool {
cmd/compile/internal/types/type.go:249:func (t *Type) HasShape() bool   { return t.flags&typeHasShape != 0 }
cmd/compile/internal/types/type.go:1403:func (t *Type) HasNil() bool {
cmd/compile/internal/types/type.go:1607:func (t *Type) HasPointers() bool {
cmd/internal/dwarf/dwarf.go:956:func HasChildren(die *DWDie) bool {
cmd/link/internal/loader/loader.go:124:func (bm Bitmap) Has(i Sym) bool {
crypto/x509/pkix/pkix.go:295:func (certList *CertificateList) HasExpired(now time.Time) bool {
vendor/golang.org/x/sys/cpu/cpu_s390x.go:74:func (s *facilityList) Has(fs ...facility) bool {
vendor/golang.org/x/sys/cpu/cpu_s390x.go:118:func (q *queryResult) Has(fns ...function) bool {
go/types/api.go:429:func (tv TypeAndValue) HasOk() bool {
net/http/h2_bundle.go:1472:func (f http2Flags) Has(v http2Flags) bool {
net/http/h2_bundle.go:2192:func (f *http2SettingsFrame) HasDuplicates() bool {
net/http/h2_bundle.go:2427:func (f *http2HeadersFrame) HasPriority() bool {
net/url/url.go:948:func (v Values) Has(key string) bool {
path/filepath/path_windows.go:13:func HasPrefix(p, prefix string) bool {
path/filepath/path_plan9.go:15:func HasPrefix(p, prefix string) bool {
path/filepath/path_unix.go:17:func HasPrefix(p, prefix string) bool {
database/sql/fakedb_test.go:1157:func (rc *rowsCursor) HasNextResultSet() bool {
internal/testenv/testenv.go:47:func HasGoBuild() bool {
internal/testenv/testenv.go:129:func HasGoRun() bool {
internal/testenv/testenv.go:145:func HasParallelism() bool {
internal/testenv/testenv.go:294:func HasSrc() bool {
internal/testenv/testenv.go:304:func HasExternalNetwork() bool {
internal/testenv/testenv.go:323:func HasCGO() bool {
internal/testenv/testenv.go:392:func HasSymlink() bool {
internal/testenv/testenv.go:407:func HasLink() bool {
internal/stringslite/strings.go:16:func HasPrefix(s, prefix string) bool {
internal/stringslite/strings.go:20:func HasSuffix(s, suffix string) bool {
internal/abi/type.go:182:func (t *Type) HasName() bool {
internal/abi/type.go:629:func (n Name) HasTag() bool {
internal/syscall/unix/nonblocking_unix.go:19:func HasNonblockFlag(flag int) bool {
internal/syscall/unix/nonblocking_js.go:13:func HasNonblockFlag(flag int) bool {
internal/syscall/unix/nonblocking_wasip1.go:22:func HasNonblockFlag(flag int) bool {
internal/cpu/cpu_s390x.go:63:func (q *queryResult) Has(fns ...function) bool {
internal/cpu/cpu_s390x.go:114:func (s *facilityList) Has(fs ...facility) bool {
internal/dag/parse.go:79:func (g *Graph) HasEdge(from, to string) bool {
internal/profile/profile.go:457:func (p *Profile) HasFunctions() bool {
internal/profile/profile.go:468:func (p *Profile) HasFileLines() bool {
rawr@carbonite: ~/Projects/go/src (master) 

$ rgrep "func.* Present[(A-Z]"
cmd/go/internal/load/flag.go:83:func (f *PerPackageFlag) Present() bool {
```

Reply:
`Has` makes it sound like a container/set type more than it is. (I guess it's a set of 0 or 1 things, but that's a stretch)

`Present` is widely used? I can't think of ever seeing it? In std, there's only `net/mail.ErrHeaderNotPresent`. Nothing in corp.  In oss, only `func (t *strideTable[T]) delete(addr uint8, prefixLen int) (wasPresent bool) {`.

How about `IsSet`?

Reply:
> `Has` makes it sound like a container/set type more than it is.

In my mind, it is a container type?

Reply:
Yeah, but Has is a transitive verb. "Has what?" I don't see any `Has() bool` in your list above. They're all `HasObject() bool` or `Has(object) bool`

Reply:
Renamed to `IsSet`.

Reply:
SGTM.